lens/private/list/flatten.rkt

```diff
@@ -0,0 +1,204 @@
+     (define-values [f-lst f-rst]
+       (split-at flattened (length s-lst)))
+     (cons f-lst (restore-structure s-rst f-rst))]))
+
+;; append*-lens : (Lens (Listof (Listof Any)) (Listof Any))
+;; where the only valid views are lists with the same length as the
+;; result of applying append* to the target.
+;; This is conceptually equivalent to (flatten/depth-lens 2)
+(define append*-lens
+  (make-lens
+   append*
+   restore-structure))
+
+;; (define-type (Listof* A n)
+;;   (cond [(zero? n) A]
+;;         [else (Listof* (Listof A) (sub1 n))]))
+
+;; flatten/depth-lens : (Lens (Listof* Any n) (Listof Any))
+;; where the only valid views are lists with the same length as the
+;; result of (flatten/depth n target)
+(define (flatten/depth-lens n)
+  (make-lens
+   (flatten/depth n _)
+   (unflatten/depth n _ _)))
+
+;; flatten/depth : n (Listof* A n) -> (Listof A)
+(define (flatten/depth n lst*)
+  (cond [(zero? n) (list lst*)]
+        [else (append*n (sub1 n) lst*)]))
+
+;; unflatten/depth : n (Listof* A n) (Listof B) -> (Listof* B n)
+(define (unflatten/depth n lst* lst)
+  (cond [(zero? n)
+         (match-define (list v) lst)
+         v]
+        [else
+         (unappend*n (sub1 n) lst* lst)]))
+
+;; append*n : n (Listof (Listof* A n)) -> (Listof A)
+(define (append*n n lst*)
+  (cond [(zero? n) lst*]
+        [else (append*n (sub1 n) (append* lst*))]))
+
+;; unappend*n : n (Listof (Listof* A n)) (Listof B) -> (Listof (Listof* B n))
+(define (unappend*n n lst* lst)
+  (cond [(zero? n) lst]
+        [else (restore-structure
+               lst*
+               (unappend*n (sub1 n) (append* lst*) lst))]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(module+ test
+  (test-case "append*-lens"
+    (check-equal? (lens-view append*-lens (list (list 1) (list 2 3) (list)))
+                  (list 1 2 3))
+    (check-equal? (lens-set append*-lens
+                            (list (list 1) (list 2 3) (list))
+                            (list 'a 'b 'c))
+                  (list (list 'a) (list 'b 'c) (list)))
+
+    (check-equal? (lens-transform append*-lens
+                                  (list (list 1) (list 2 3) (list))
+                                  (lambda (lst)
+                                    ;; a length-preserving computation
+                                    (let loop ([acc (list)] [sum 0] [lst lst])
+                                      (match lst
+                                        [(list) (reverse acc)]
+                                        [(cons fst rst)
+                                         (loop (cons (+ sum fst) acc)
+                                               (+ sum fst)
+                                               rst)]))))
+                  (list (list 1) (list 3 6) (list)))
+
+    (check-equal? (lens-transform append*-lens
+                                  (list (list #'(+ a)) (list #'(- a b) #'(* c d)) (list))
```

Comment:
Oh, this is here because this was my original use case for the `flatten/depth-lens`. I should probably just remove this test.